repo2docker/app.py

```diff
@@ -395,26 +408,29 @@ def fetch(self, url, ref, checkout_path):
                 "No matching content provider found for " "{url}.".format(url=url)
             )
 
+        swh_token = self.config.get("swh_token", self.swh_token)
+        if swh_token and isinstance(picked_content_provider, contentproviders.Swhid):
+            picked_content_provider.set_auth_token(swh_token)
+
         for log_line in picked_content_provider.fetch(
             spec, checkout_path, yield_output=self.json_logs
         ):
             self.log.info(log_line, extra=dict(phase="fetching"))
 
         if not self.output_image_spec:
-            self.output_image_spec = (
-                "r2d" + escapism.escape(self.repo, escape_char="-").lower()
-            )
+            image_spec = "r2d" + self.repo
             # if we are building from a subdirectory include that in the
             # image name so we can tell builds from different sub-directories
             # apart.
             if self.subdir:
-                self.output_image_spec += escapism.escape(
-                    self.subdir, escape_char="-"
-                ).lower()
+                image_spec += self.subdir
             if picked_content_provider.content_id is not None:
-                self.output_image_spec += picked_content_provider.content_id
+                image_spec += picked_content_provider.content_id
             else:
-                self.output_image_spec += str(int(time.time()))
+                image_spec += str(int(time.time()))
+            self.output_image_spec = escapism.escape(
+                image_spec, escape_char="-"
+            ).lower()
```

Comment:
What is the thinking behind moving the escaping to the end? If I remember correctly the escaping was done before adding the content ID and the time stamp because we know those should be "safe" and hence we get less "ugly escaping"? But I am unsure.

This means if we spend time now to think about moving the escaping (or not) we could add a comment about why it is in the position it is so that future generations don't have to guess :D

Reply:
well it's documented in the git commit message: you have no guarantee what the content_id will be, so it needs to be escaped as well. Otherwise you may have hard-to-understand docker errors (typically when the content_id contains several ":" characters, because the generated image name is then invalid for docker.

Reply:
> [...] content ID and the time stamp because we know those should be "safe" 

To be clear, it is this assertion I do not agree with. There is no restriction on what the content_id can be. The content_id docstring of the `ContentProvider` base class states:

>         The type content ID can be any string.

For example in the SWHID content provider I naturally used the swhid itself as content_id, but since this later consists in 4 parts separated by columns (:), it makes repo2docker fail at giving the docker image a valid name.

I could change the code of the swhid provider, but it looks to me that ensuring the output image name to be valid is best to be done here than distribute the responsibility in all content providers.

Reply:
Sounds like a good reasoning. I didn't have the content ID comment/docstring in my head any more.